wp-content/plugins/goonj-blocks/build/functions.php

```diff
@@ -15,4 +15,208 @@ function gb_format_time_range($start_date, $end_date, $format = 'h:i A') {
+            $contactCity = isset($contactData['address_primary.city']) ? strtolower($contactData['address_primary.city']) : '';
+            if (in_array($contactCity, ['patna', 'ranchi', 'bhubaneshwar'])) {
+                // generate physical slot for these cities
+                return generate_slots($assignedOfficeId, $defaultMaxSlot, $physicalInductionType, $inductionSlotStartDate);
+            }
+            // generate online induction slots for other cities
+            return generate_slots($assignedOfficeId, $defaultMaxSlot, $onlineInductionType, $inductionSlotStartDate);
+        }
+
+        // Determine if a Goonj office exists in the contact's state and schedule accordingly
+        $officeContact = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('id', 'display_name')
+            ->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+            ->addWhere('address_primary.state_province_id', '=', $contactStateId)
+            ->execute();
+
+        if ($officeContact->count() === 0) {
+            // generate online induction slots for state having no office
+            return generate_slots($assignedOfficeId, $defaultMaxSlot, $onlineInductionType, $inductionSlotStartDate);
+        }
+        // generate physical induction slots having office in their states
+        return generate_slots($assignedOfficeId, $defaultMaxSlot, $physicalInductionType, $inductionSlotStartDate);
+    } catch (\Exception $e) {
+        \Civi::log()->error('Error generating induction slots', [
+            'contactId' => $contactId,
+            'error' => $e->getMessage()
+        ]);
+        return [];
+    }
+}
+
+/**
+ * Generate specific induction slots based on induction type and scheduled activities.
+ *
+ * @param int $assignedOfficeId The office ID where slots are generated.
+ * @param int $maxSlots The maximum number of slots to generate.
+ * @param array $scheduledActivities List of already scheduled induction activities.
+ * @param string $inductionType The type of induction ('Processing_Unit' or 'Online_only_selected_by_Urban_P').
+ * @param DateTime $startDate The start date for generating slots.
+ * @return array Array of generated slots with day, date, time, count, and type.
+ */
+function generate_slots($assignedOfficeId, $maxSlots, $inductionType, $startDate) {
+    $slots = [];
+    $slotCount = 0;
+    $highActivityCountDays = 0;
+
+    try {
+        // Fetch office induction details for induction scheduling
+        $officeDetails = \Civi\Api4\Contact::get(FALSE)
+            ->addSelect('display_name', 'Goonj_Office_Details.Physical_Induction_Slot_Days:name', 'Goonj_Office_Details.Physical_Induction_Slot_Time', 'Goonj_Office_Details.Online_Induction_Slot_Days:name', 'Goonj_Office_Details.Online_Induction_Slot_Time')
+            ->addWhere('contact_sub_type', 'CONTAINS', 'Goonj_Office')
+            ->addWhere('id', '=', $assignedOfficeId)
+            ->execute()->first();
```

Comment:
Can you please check everwhere and change it to single whenever needed

Reply:
sure